services/rmetrics/pending_events.go

```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+// NewPendingEventsRegistry creates a new PendingEventsRegistry. By default, metrics are not published to the global published metrics registry, until [Publish] is called.
+func NewPendingEventsRegistry(opts ...Option) PendingEventsRegistry {
+	per := &pendingEventsRegistry{
+		registry: metric.NewRegistry(),
+	}
+	for _, opt := range opts {
+		opt(per)
+	}
+	return per
+}
+
+type pendingEventsRegistry struct {
+	registryMu sync.RWMutex
+	published  bool
+	registry   metric.Registry
+}
+
+// IncreasePendingEvents increments three gauges, the dest & workspace-specific gauge, plus two aggregate (global) gauges
+func (pem *pendingEventsRegistry) IncreasePendingEvents(tablePrefix, workspace, destType string, value float64) {
+	pem.registryMu.RLock()
+	defer pem.registryMu.RUnlock()
+
+	pem.PendingEvents(tablePrefix, workspace, destType).Add(value)
+	pem.PendingEvents(tablePrefix, All, destType).Add(value)
+	pem.PendingEvents(tablePrefix, All, All).Add(value)
+	pem.registry.MustGetGauge(pendingEventsMeasurementAll{tablePrefix, destType}).Add(value)
+	pem.registry.MustGetGauge(pendingEventsMeasurementAll{tablePrefix, All}).Add(value)
+}
+
+// DecreasePendingEvents increments three gauges, the dest & workspace-specific gauge, plus two aggregate (global) gauges
+func (pem *pendingEventsRegistry) DecreasePendingEvents(tablePrefix, workspace, destType string, value float64) {
+	pem.registryMu.RLock()
+	defer pem.registryMu.RUnlock()
+	pem.PendingEvents(tablePrefix, workspace, destType).Sub(value)
+	pem.PendingEvents(tablePrefix, All, destType).Sub(value)
+	pem.PendingEvents(tablePrefix, All, All).Sub(value)
+	pem.registry.MustGetGauge(pendingEventsMeasurementAll{tablePrefix, destType}).Sub(value)
+	pem.registry.MustGetGauge(pendingEventsMeasurementAll{tablePrefix, All}).Sub(value)
+}
+
+// PendingEvents gets the measurement for pending events metric
+func (pem *pendingEventsRegistry) PendingEvents(tablePrefix, workspace, destType string) metric.Gauge {
+	return pem.registry.MustGetGauge(newPendingEventsMeasurement(tablePrefix, workspace, destType))
+}
+
+// Publish publishes the metrics to the global published metrics registry
+func (pem *pendingEventsRegistry) Publish() {
+	pem.registryMu.Lock()
+	defer pem.registryMu.Unlock()
+	if pem.published {
+		return
+	}
+	pem.published = true
+
+	publishedRegistry := metric.Instance.GetRegistry(metric.PublishedMetrics)
+	pem.registry.Range(func(key, value any) bool { // copy all gauge metrics to the published registry
+		m := key.(metric.Measurement)
+		switch value := value.(type) {
+		case metric.Gauge:
+			publishedRegistry.MustGetGauge(m).Set(value.Value())
+		}
+		return true
+	})
+	pem.registry = publishedRegistry
+}
+
+// Reset resets the registry to a new, non published one and clears the global published metrics registry
+func (pem *pendingEventsRegistry) Reset() {
+	pem.registryMu.Lock()
+	defer pem.registryMu.Unlock()
+	pem.registry = metric.NewRegistry()
+	metric.Instance.Reset()
```

Comment:
Do we need to set `published` to false here?
Is it safe to reset the global instance? will we lose any metrics?

Reply:
thanks for catching that!

Reply:
in the global instance only pending events metrics are stored

Reply:
Added a scenario in tests for covering publishing after reset, thanks!